File: 2646-kth-largest-sum-in-a-binary-tree/kth-largest-sum-in-a-binary-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    long long kthLargestLevelSum(TreeNode* root, int k) {
        queue<TreeNode*> q;
        priority_queue<long long, vector<long long>, greater<long long>> pq;
        q.push(root);
        while(q.size() > 0){
            int l = q.size();
            long long s = 0;
            while (l > 0){
                TreeNode* temp = q.front();
                q.pop();
                s += temp->val;
                if (temp->left != nullptr) q.push(temp->left);
                if (temp->right != nullptr) q.push(temp->right);
                l -= 1;
            }
            pq.push(s);
            if (pq.size() > k){
                pq.pop();
            }
        }

        return pq.size() < k ? -1 : pq.top();
    }
};

File: 2646-kth-largest-sum-in-a-binary-tree/kth-largest-sum-in-a-binary-tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public long kthLargestLevelSum(TreeNode root, int k) {
        Queue<TreeNode> q = new LinkedList<TreeNode>();
        PriorityQueue<Long> pq = new PriorityQueue<Long>();
        q.add(root);
        while(q.size() > 0){
            int l = q.size();
            long s = 0;
            while (l > 0){
                TreeNode temp = q.poll();
                s += temp.val;
                if (temp.left != null) q.add(temp.left);
                if (temp.right != null) q.add(temp.right);
                l -= 1;
            }
            pq.add(s);
            if (pq.size() > k){
                pq.remove();
            }
        }

        return pq.size() < k ? -1 : pq.peek();
    }
}